File: python/src/imc_tally.py
"""Tally end-of-timestep quantities."""

import numpy as np

import imc_global_mat_data as mat
import imc_global_mesh_data as mesh
import imc_global_phys_data as phys
import imc_global_time_data as time


def run():
    """Tally end-of-timestep quantities."""
    print("\n" + "-" * 79)
    print("Tally step ({:4d})".format(time.step))
    print("-" * 79)

    # Start-of-step radiation energy density
    radnrgdens = np.zeros(mesh.ncells)
    radnrgdens[:] = phys.a * mesh.temp[:] ** 4

    # Temperature increase
    nrg_inc = np.zeros(mesh.ncells)
    nrg_inc[:] = (
        mesh.nrgdep[:] / mesh.dx
        - mesh.fleck[:] * phys.c * time.dt * mesh.sigma_p[:] * radnrgdens[:]
    )

    mesh.temp[:] = mesh.temp[:] + nrg_inc[:] / mat.bee

    print("\nMesh temperature:")
    print(mesh.temp)

File: fortran/src/imc_tally.f90
subroutine imc_tally()

    use imc_global_mat_data,  only: bee
    use imc_global_mesh_data, only: ncells, dx, temp, fleck, sigma_p, nrgdep
    use imc_global_phys_data, only: c, a
    use imc_global_time_data, only: step, dt

    implicit none

    real(kind=8), dimension(ncells) :: radnrgdens
    real(kind=8), dimension(ncells) :: nrg_inc

    write(*,'(/,A)')    '-------------------------------------------------------------------------------'
    write(*,'(A,I4,A)') 'Tally step (',step,')'
    write(*,'(A)')      '-------------------------------------------------------------------------------'

    ! Start-of-step radiation energy density
    radnrgdens(:) = a * temp(:)**4

    ! Temperature increase
    nrg_inc(:)  = ( &
        nrgdep(:) / dx &
        - fleck(:) * c * dt * sigma_p(:) * radnrgdens(:) &
    )

    temp(:) = temp(:) + nrg_inc(:) / bee

    write(*,'(/,A)') 'Mesh temperature:'
    write(*,'(4(ES16.8))') temp

end subroutine imc_tally
